### ISSUE TITLE
Memory leak for QF_LIA instances (parsing and incremental solving)

### ISSUE THREAD

**Issue by rainoftime** [usi-verification-and-security/opensmt#95]
Hi, for the following formula,
~~~~
(and (or (<= 4 y) (<= 4 x)) (or (<= 3 x) (and (<= 4 x) (and (<= 4 x) (or (<= 4 y) (<= 4 x)))))))
~~~~

opensmt (commit a492c61) throws a memory leak
~~~~
==13207==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x7fadaebb8592 in operator new(unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x99592)
    #1 0x50f660 in smt2newparse(Smt2newContext*) /home/rainoftime/Work/tofuzz/opensmt/build/src/parsers/smt2new/smt2newparser.yy:97
    #2 0x4b3038 in Interpret::interpFile(_IO_FILE*) /home/rainoftime/Work/tofuzz/opensmt/src/api/Interpret.cc:1052
    #3 0x4a5869 in main /home/rainoftime/Work/tofuzz/opensmt/src/bin/opensmt.cc:129
    #4 0x7fadad58882f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Direct leak of 4 byte(s) in 1 object(s) allocated from:
    #0 0x7fadaeb8134f in strdup (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x6234f)
    #1 0x51be1e in smt2newlex(YYSTYPE*, YYLTYPE*, void*) /home/rainoftime/Work/tofuzz/opensmt/build/src/parsers/smt2new/smt2newlexer.ll:136
    #2 0x50ef6b in smt2newparse(Smt2newContext*) /home/rainoftime/Work/tofuzz/opensmt/build/src/parsers/smt2new/smt2newparser.cc:1558
    #3 0x4b3038 in Interpret::interpFile(_IO_FILE*) /home/rainoftime/Work/tofuzz/opensmt/src/api/Interpret.cc:1052
    #4 0x4a5869 in main /home/rainoftime/Work/tofuzz/opensmt/src/bin/opensmt.cc:129
    #5 0x7fadad58882f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: 28 byte(s) leaked in 2 allocation(s).
~~~~

**Comment by blishko** [comment]
I am not seeing this error, can you please attach the whole SMT-LIB script?

**Comment by rainoftime** [comment]
@blishko  The above line is the whole input. 
I encountered the issue "accidentally".

**Comment by rainoftime** [comment]
There is a parsing error 
At line 1: syntax error, unexpected TK_SYM

**Comment by blishko** [comment]
Ah, OK. I understand.
At the moment we are focusing on fixing errors encountered for valid inputs.
I personally don't see a memory leak for an invalid input as a problem, since the solver will not be doing anything for such input.

**Comment by rainoftime** [comment]
@blishko  Can you check this?

For the following formula,
~~~~
(set-logic QF_LIA)
(declare-fun i0 () Int)
(push 1)
(pop 1)
(assert (= (+ 94 i0) i0))
(push 1)
(push 1)
(check-sat)
~~~~
cvc4 throws a leak
~~~~
==60523==ERROR: LeakSanitizer: detected memory leaks                                                                                    [29/36]

Indirect leak of 88 byte(s) in 1 object(s) allocated from:
    #0 0x7f11d0ab2498 in operator new(unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xe0498)
    #1 0x5c3bb2 in __gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<PTRef const, FastRational> > >::allocate(unsigned long, void const*) /
usr/include/c++/7/ext/new_allocator.h:111
    #2 0x5c3965 in std::allocator_traits<std::allocator<std::_Rb_tree_node<std::pair<PTRef const, FastRational> > > >::allocate(std::allocator<
std::_Rb_tree_node<std::pair<PTRef const, FastRational> > >&, unsigned long) /usr/include/c++/7/bits/alloc_traits.h:436
    #3 0x5c32ce in std::_Rb_tree<PTRef, std::pair<PTRef const, FastRational>, std::_Select1st<std::pair<PTRef const, FastRational> >, std::less
<PTRef>, std::allocator<std::pair<PTRef const, FastRational> > >::_M_get_node() /usr/include/c++/7/bits/stl_tree.h:588
    #4 0x65e686 in std::_Rb_tree_node<std::pair<PTRef const, FastRational> >* std::_Rb_tree<PTRef, std::pair<PTRef const, FastRational>, std::_
Select1st<std::pair<PTRef const, FastRational> >, std::less<PTRef>, std::allocator<std::pair<PTRef const, FastRational> > >::_M_create_node<std
::piecewise_construct_t const&, std::tuple<PTRef const&>, std::tuple<> >(std::piecewise_construct_t const&, std::tuple<PTRef const&>&&, std::tu
ple<>&&) /usr/include/c++/7/bits/stl_tree.h:642
    #5 0x65d314 in std::_Rb_tree_iterator<std::pair<PTRef const, FastRational> > std::_Rb_tree<PTRef, std::pair<PTRef const, FastRational>, std
::_Select1st<std::pair<PTRef const, FastRational> >, std::less<PTRef>, std::allocator<std::pair<PTRef const, FastRational> > >::_M_emplace_hint
_unique<std::piecewise_construct_t const&, std::tuple<PTRef const&>, std::tuple<> >(std::_Rb_tree_const_iterator<std::pair<PTRef const, FastRat
ional> >, std::piecewise_construct_t const&, std::tuple<PTRef const&>&&, std::tuple<>&&) /usr/include/c++/7/bits/stl_tree.h:2398
    #6 0x65c73c in std::map<PTRef, FastRational, std::less<PTRef>, std::allocator<std::pair<PTRef const, FastRational> > >::operator[](PTRef co
nst&) /usr/include/c++/7/bits/stl_map.h:493
    #7 0x655817 in LAExpression::initialize(PTRef, bool) /home/rainoftime/Work/tofuzz/opensmt/src/simplifiers/LA.cc:98
    #8 0x5bb7c1 in LAExpression::LAExpression(LALogic&, PTRef, bool) /home/rainoftime/Work/tofuzz/opensmt/src/simplifiers/LA.h:53
    #9 0x5bb836 in LAExpression::LAExpression(LALogic&, PTRef) /home/rainoftime/Work/tofuzz/opensmt/src/simplifiers/LA.h:56
    #10 0x5a26f8 in LALogic::arithmeticElimination(vec<PTRef> const&, Map<PTRef, PtAsgn, PTRefHash, Equal<PTRef> >&) /home/rainoftime/Work/tofu
zz/opensmt/src/logics/LALogic.cc:119
    #11 0x5a3d3e in LALogic::retrieveSubstitutions(vec<PtAsgn> const&, Map<PTRef, PtAsgn, PTRefHash, Equal<PTRef> >&) /home/rainoftime/Work/tof
uzz/opensmt/src/logics/LALogic.cc:233
    #12 0x5c6365 in Theory::computeSubstitutions(PTRef, vec<PFRef> const&, int) /home/rainoftime/Work/tofuzz/opensmt/src/logics/Theory.cc:129
    #13 0x675167 in LIATheory::simplify(vec<PFRef> const&, int) /home/rainoftime/Work/tofuzz/opensmt/src/logics/LIATheory.cc:21
    #14 0x4c1255 in MainSolver::simplifyFormulas(char**) /home/rainoftime/Work/tofuzz/opensmt/src/api/MainSolver.cc:121
~~~~

**Comment by blishko** [comment]
Yes, that's a memory leak. Thanks!

**Comment by blishko** [comment]
Fixed in #147
